Tutorials
=========


------------

Single cell ATAC-seq
--------------------

To get started, we recommend epiScanpy's analysis pipeline for scATAC-seq data for 10k PBMCs from 10x Genomics. This tutorial focuses on preprocessing, clustering, identification of cell types via known marker genes and trajectory inference. The tutorial can be found here.


.. raw:: html
   <img src="http://github.com/DaneseAnna/episcanpy-pictures/blob/master/pbmc_PCA_coverage.png" style="width: 100px">
   <img src="http://github.com/DaneseAnna/episcanpy-pictures/blob/master/pbmc_PCA_coverage.png" style="width: 100px">
   
Soon available, there will be a tutorial on preprocessing, clustering and identification of cell types for the Cusanovich mouse scATAC-seq atlas (ref Cusanovich mouse atlas paper) (prefrontal cortex data). In this tutorial we focus on the use of different feature space count matrices (peak and enhancer based count matrices).





------------

Single cell DNA methylation
---------------------------

Here you can find a tutorial for the preprocessing, clustering and identification of cell types for single-cell DNA methylation data using the publicly available data from Luo et al. (Ecker lab). The tutorial shows how to build the count matrices for the different feature spaces and how to use them and compare the results. The data used comes from mouse brain (frontal cortex).

.. image:: "http://drive.google.com/open?id=1TbCZAtglylm0k_Ss_JNMGGCbBNKrYKcs"
   :width: 400

.. raw:: html

   <img src="http://drive.google.com/open?id=1TbCZAtglylm0k_Ss_JNMGGCbBNKrYKcs" style="width: 100px">
   
