Tutorials
=========


------------

single cell ATAC
----------------

For getting started, we recommend `epiScanpy's adaptation PBMC clustering tutorial of `Scanpy's to the 10k PBMCs from 10x Genomics. The tutorial focus on preprocessing, clustering, identification of cell types via known marker genes and trajectory inference. tutorial here_ 



.. raw:: html

   <img src="http://falexwolf.de/img/scanpy_usage/170505_seurat/filter_genes_dispersion.png" style="width: 100px"><img 
   
Soon available, there will be a tutorial on clustering comparison between peak and enhancer based count matrices usinf prefrontal cortex data from the Cusanovich mouse ATAC atlas (ref Cusanovich mouse atlas paper).

.. raw:: html

   <img src="http://falexwolf.de/img/scanpy_usage/170505_seurat/filter_genes_dispersion.png" style="width: 100px"><img 

------------

Single cell methylation
-----------------------

For trajectory inference on complex datasets, we offer several examples `here <https://github.com/theislab/paga>`__. Get started `here <https://nbviewer.jupyter.org/github/theislab/paga/blob/master/blood/paul15/paul15.ipynb>`__ for the following result on hematopoiesis.

.. raw:: html

   <img src="http://www.falexwolf.de/img/paga_paul15.png" style="width: 450px">
   
   
.. _tutorialhere: https://github.com/colomemaria/epiScanpy/blob/readthedocs/docs/beta_tutorial_10x_pbmc.html
