Tutorials
=========


------------

single cell ATAC
----------------

For getting started, we recommend epiScanpy's adaptation PBMC clustering tutorial of `Scanpy's to the 10k PBMCs from 10x Genomics. The tutorial focus on preprocessing, clustering, identification of cell types via known marker genes and trajectory inference. tutorial here_ 

.. raw:: html

   <img src="https://github.com/colomemaria/epiScanpy/blob/master/pbmc_PCA_coverage.png" style="width: 100px"><img
   <img src="https://github.com/colomemaria/epiScanpy/blob/master/pbmc_umap.png" style="width: 100px"><img
   <img src="https://github.com/colomemaria/epiScanpy/blob/master/pbmc_diffmap.png" style="width: 100px"><img
   
   
Soon available, there will be a tutorial on clustering comparison between peak and enhancer based count matrices usinf prefrontal cortex data from the Cusanovich mouse ATAC atlas (ref Cusanovich mouse atlas paper).

.. raw:: html

   <img src="https://github.com/colomemaria/epiScanpy/blob/readthedocs/docs/umap.png" style="width: 100px"><img
   <img src="https://github.com/colomemaria/epiScanpy/blob/readthedocs/docs/heatmap.png" style="width: 100px"><img



------------

Single cell methylation
-----------------------

Soon a tutorial on windows based methylation data analysis using the publicly available data from Luo et al. (Ecker lab).
Another tutorial will arrive as a follow up comparing promoters and enhancers in frontal cortex brain.

.. raw:: html

   <img src="https://github.com/colomemaria/epiScanpy/blob/master/umap_markers_hodology_ecker.png" style="width: 450px">
   
   
.. _tutorialhere: https://github.com/colomemaria/epiScanpy/blob/readthedocs/docs/beta_tutorial_10x_pbmc.html
